### ISSUE TITLE
Typescript 2.9 breaks: Error TS2536: Type 'P' cannot be used to index type 'Si' 

### ISSUE THREAD

**Issue by lolgesten** [cyclejs/cyclejs#815]
**Code to reproduce the issue:**

Use `@cycle/run` package with typescript ^2.9

**Actual behavior:**

```
[12:44:56] [bro] TypeScript error
TypeScript error: node_modules/@cycle/run/lib/cjs/types.d.ts(32,36): Error TS2536: Type 'P' cannot be used to index type 'Si'.
[12:44:56] [bro] TypeScript error
TypeScript error: node_modules/@cycle/run/lib/cjs/types.d.ts(32,43): Error TS2536: Type 'P' cannot be used to index type 'So'.
```

**Versions of packages used:**

    "@cycle/dom": "^20.4.0",
    "@cycle/isolate": "^3.3.0",
    "@cycle/run": "^4.1.0",

**Comment by staltz** [comment]
TypeScript once in a while issues breaking changes under the minor number (e.g. 2.**9**.1) which is against SemVer. We will aim to update cycle/run with TS 2.9, but note that you are using an incompatible TS. Prefer using `~2.8.0` not `^2.8.0` as the version in your dependencies.

**Comment by lolgesten** [comment]
I never actually saw the `Drivers` type until now that it bit me. My code looks like this, and I never reflected on what that `run` arg actually was... 🤷‍♀️ 

```typescript
const drivers = {
    ChromeDriver,
    MediaDriver,
    RecorderDriver: makeRecorderDriver(api),
    WongDriver,
};

// tslint:disable-next-line:no-expression-statement
run(main, drivers);
```

**Comment by sledorze** [comment]
@staltz I came to your issue from twitter.

May that fix it?
```
[P in keyof (So & Si)]
```
Replace by:
```
[P in (keyof So) & (keyof Si)]
```

**Comment by staltz** [comment]
Good fix! I think that did the trick. PR coming soon

**Comment by staltz** [comment]
I'm not sure the suggestion helped.

As far as I understand, `keyof (So & Si)` picks all the keys that exist in either So or Si. `(keyof So) & (keyof Si)` picks only keys of So that are also keys of Si. `keyof T` returns a union type of strings. And `unionTypeOfStringsA & unionTypeOfStringsB` is basically the intersection of those.

**Comment by sledorze** [comment]
```typescript

interface So {
  a: string
  b: string
}

interface Si {
  b: string
  c: string
}

So & Si  // interface {  a: string, b: string, c: string  }

keyof (So & Si) // 'a' | 'b' | 'c'

keyof So // 'a' | 'b'
keyof Si // 'b' | 'c'

(keyof So) & (keyof Si) // ('a' | 'b') & ('b' | 'c')   -->  'b'

```

**Comment by infctr** [comment]
Is the package with fixed typings going to be published any time soon? Just checked `4.4.0` and it doesn't have the changes yet

**Comment by jvanbruegge** [comment]
I finished this up, it's currently waiting for a review and release
